File: 0876-middle-of-the-linked-list/0876-middle-of-the-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        //optimal method
        if(!head) return head;
        if(!head->next) return head;
        if(!head->next->next) return head->next;
        
        
        ListNode * slow = head;
        ListNode * fast = head->next;
        
        
        
        while(fast){
            fast = fast->next;
            if(fast != NULL) fast = fast->next;
            slow = slow->next;
        }
        
        return slow;
        
    }
};

File: 0876-middle-of-the-linked-list/0876-middle-of-the-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if(head==None): 
            return head
        if(head.next==None):
            return head
        if(head.next.next == None):
            return head.next
        fast = head.next
        slow = head
        while(fast):
            fast = fast.next
            if(fast):
                fast = fast.next
            slow = slow.next
        return slow